src/structs/probability_tables.rs

```diff
@@ -104,112 +114,100 @@ impl ProbabilityTables {
+                for i in 0..49 {
+                    best_prior[i] = left[i].abs();
+                }
+            } else if self.above_present {
+                let above = block_context.above(image_data).get_block();
+                for i in 0..49 {
+                    best_prior[i] = above[i].abs();
+                }
+            }
         }
 
-        return ProbabilityTablesCoefficientContext {
-            best_prior: best_prior as i32,
-            num_non_zeros_bin: ProbabilityTables::num_non_zeros_to_bin(num_nonzeros_left),
-            best_prior_bit_len: u32_bit_length(cmp::min(
-                best_prior, /* no need for abs here as in original source */
-                1023,
-            ))
-            .into(),
-        };
+        return best_prior;
     }
 
+    #[inline(always)]
     pub fn calc_coefficient_context8_lak<const ALL_PRESENT: bool, const HORIZONTAL: bool>(
         &self,
-        image_data: &BlockBasedImage,
         qt: &QuantizationTables,
         coefficient: usize,
-        block_context: &BlockContext,
+        here: &[i16; 64],
+        above: &[i16; 64],
+        left: &[i16; 64],
         num_non_zeros_x: u8,
     ) -> ProbabilityTablesCoefficientContext {
-        let coef_idct;
-
         let mut compute_lak_coeffs_x: [i32; 8] = [0; 8];
         let mut compute_lak_coeffs_a: [i32; 8] = [0; 8];
 
         debug_assert_eq!(HORIZONTAL, (coefficient & 7) != 0);
 
-        let coef_idct_offset;
+        let coef_idct;
         if HORIZONTAL && (ALL_PRESENT || self.above_present) {
             assert!(coefficient < 8); // avoid bounds check later
 
             // y == 0: we're the x
-            let here = block_context.here(image_data);
-            let above = block_context.above(image_data);
-
             // the compiler is smart enough to unroll this loop and merge it with the subsequent loop
             // so no need to complicate the code by doing anything manual
 
             for i in 0..8 {
-                let cur_coef = RASTER_TO_ALIGNED[(coefficient + (i * 8)) as usize] as usize;
-
-                compute_lak_coeffs_x[i] = if i != 0 {
-                    here.get_coefficient(cur_coef).into()
-                } else {
-                    0
-                };
-                compute_lak_coeffs_a[i] = above.get_coefficient(cur_coef).into();
+                let cur_coef = usize::from(RASTER_TO_ALIGNED[usize::from(coefficient + (i * 8))]);
+
+                let sign = if (i & 1) != 0 { -1 } else { 1 };
+
+                compute_lak_coeffs_x[i] = if i != 0 { here[cur_coef].into() } else { 0 };
+                compute_lak_coeffs_a[i] = (sign * above[cur_coef]).into();
             }
 
-            coef_idct = qt.get_icos_idct_edge8192_dequantized_x();
-            coef_idct_offset = coefficient * 8;
+            coef_idct =
+                &qt.get_icos_idct_edge8192_dequantized_x()[coefficient * 8..(coefficient + 1) * 8];
```

Comment:
This pattern has bitten me a bunch before: you get some really odd behavior
first it checks if (coefficient * 8) < (coefficient + 1) * 8
secondly it does two bounds checks...

couple of ways to improve this
qt.get_icos_idct_edge8192_dequantized_x().splt_at(coefficient * 8).1.split_at(8);

alternative approach is to make your coefficient a u32 and then upcast it to u64 so that the compiler does not have to check for wrapping.

It's possible that the compiler can prove the values are less than the table size.

Another approach you could go with here is to make coefficient a u8 and make your qt.get_icos_idct_edge8192_dequantized_x() a 256-big vector...so that the compiler can be right sure that it's in-bounds and can elide all checks.

Reply:
ya so recommend `icos_idct_edge8192_dequantized_x: [i32; 64]` into  `icos_idct_edge8192_dequantized_x: [i32; 256]`
Then change 
` pub fn get_icos_idct_edge8192_dequantized_x(&self) -> &[i32;256] {`